pandas/core/dtypes/missing.py

```diff
@@ -63,7 +79,27 @@
 _dtype_str = np.dtype(str)
 
 
-def isna(obj):
+@overload
+def isna(obj: Scalar) -> bool:
+    ...
+
+
+@overload
+def isna(obj: Index | ExtensionArray | np.ndarray | list) -> npt.NDArray[np.bool_]:
+    ...
+
+
+@overload
+def isna(obj: NDFrameT) -> NDFrameT:
+    ...
+
+
+@overload
+def isna(obj: object) -> bool | npt.NDArray[np.bool_] | NDFrame:
```

Comment:
isna accepts anything. If it isn't one of the already mentioned types, it always returns bool (actually Literal[False]). There is no way to have a "fallback" overload with an incompatible return type: need to return all possible types.

Reply:
Was just checking to see whether Microsoft/Virtus lab found a way: As far as I see `isna` isn't typed in the Microsoft stubs (`missing.pyi` is "missing" ;) ) and virtus lab has no [helpful](https://github.com/VirtusLab/pandas-stubs/blob/bc75133debedd69de26f1866923c0e1f2eaf16f0/third_party/3/pandas/core/dtypes/missing.pyi#L10) annotations for `isna`

Reply:
Well, `missing.pyi` is there when they ship them with VS Code.  But they didn't figure it out either.

It would be nice to be able to have an overload that says "match anything except what you matched before", e.g., instead of
```python
@overload
def isna(obj: object) -> bool | npt.NDArray[np.bool_] | NDFrame:
```
have:
```python
@overload
def isna(obj: AnythingElse) -> bool:
```
but the python typing gods would need to add that.

Reply:
> but the python typing gods would need to add that.

There's a thread about this on the SIG where it may be useful to weigh in: https://mail.python.org/archives/list/typing-sig@python.org/message/HI5RVT7ZKKQQO6J6ZASEEA3PSCKJAWEH/

Reply:
I think this is more of a bug in `mypy`.  Consider the following example:
```python
from typing import overload, Literal


class Monster:
    pass


@overload
def isMonster(x: Monster) -> bool:
    ...


@overload
def isMonster(x: list) -> list:
    ...


@overload
def isMonster(x: object) -> Literal[False]:
    ...


def isMonster(x: object) -> list | bool:
    if isinstance(x, list):
        return [isMonster(y) for y in x]
    elif isinstance(x, Monster):
        return True
    else:
        return False


reveal_type(isMonster(3))
reveal_type(isMonster(Monster()))
reveal_type(isMonster([Monster(), 5, Monster()]))
```
`pyright` with defaults doesn't complain about this.  But `mypy` does:
```
anythingelse.py:10: error: Overloaded function signatures 1 and 3 overlap with incompatible return types
anythingelse.py:15: error: Overloaded function signatures 2 and 3 overlap with incompatible return types
anythingelse.py:33: note: Revealed type is "Literal[False]"
anythingelse.py:34: note: Revealed type is "builtins.bool"
anythingelse.py:35: note: Revealed type is "builtins.list[Any]"
```
But if you add `#type: ignore[misc]` on the first 2 overloads, `mypy` no longer complains.

So for this PR, we could use this:
```python
@overload
def isna(obj: Scalar) -> bool:
    ...


@overload
def isna(  # type: ignore[misc]
    obj: ExtensionArray | np.ndarray | list,
) -> npt.NDArray[np.bool_]:
    ...


@overload
def isna(obj: NDFrameT) -> NDFrameT:  # type: ignore[misc]
    ...


@overload
def isna(  # type: ignore[misc]
    obj: IndexOpsMixin,
) -> npt.NDArray[np.bool_]:
    ...


@overload
def isna(obj: object) -> bool:
    ...


def isna(obj: object) -> bool | npt.NDArray[np.bool_] | NDFrame:
```

Semantically, that's what we want - if there is no match in the first 4 overloads, return a `bool` .

Note that `mypy` is inferring the correct types based on what was written, but is just complaining about the overloads.  I've seen `numpy` use a lot of `# type: ignore[misc]` on their overloads, so I think we should be OK in doing the same.

Reply:
@Dr-Irv I like this approach (and it could be helpful in quite a few places!) but it doesn't handle Unions well:

```py
def foo() -> Union[pd.Index, np.ndarray]:
    ...

reveal_type(pd.isna(foo()))
```
pyright and mypy infer it as `bool`, even though both `Index` and `np.ndarray` by themselves would return a `npt.NDArray[np.bool_]`

This might not happen often in user code, but there are a few places in the pandas codebase itself.

Reply:
> This might not happen often in user code, but there are a few places in the pandas codebase itself.

I tried it on your PR, and only had to change the `#type: ignore` in `ewm.py`, line 396 to be `attr-defined`.   So not sure where the "few places in the pandas codebase" are.

I believe you could also be explicit about that particular `Union` by adding this overload as the last one before the one for `isna(obj:  object)`:
```python
@overload  # type: ignore[misc]
def isna(obj: Index | np.ndarray) -> npt.NDArray[np.bool_]:
    ...
```

I don't think that's necessary since `mypy` seems to be happy without it.

Reply:
You are right, not sure what I messed up when I tried it.

I will leave this in a separate commit so it is easy to revert - I think this needs first some discussion, I see the benefit [1] but it is considered unsupported/invalid and it doesn't handle unions well. @simonjayhawkins 

[1] In this case it probably doesn't help much (not sure how often people call pd.isna with random objects) but in #44896 it would be very helpful - None and Hashable overlap

Reply:
The above union example would work, if `Index` would be added back to the overload `ArrayLike | list`. As soon as a union with types that are not covered by a single overload, pyright/mypy seem to pick the "fallback" overload.

Reply:
I was looking whether typeshed uses fallback-overloads with an incompatible return type:

They use "fallback overloads" for example for `pow` but the fallback has (unlike in our case) a wider type:
https://github.com/python/typeshed/blob/f4ae363b564de69bfad4b4346204a25a99cd6642/stdlib/builtins.pyi#L241
```py
    @overload
    def __pow__(self, __x: Literal[0], __modulo: None = ...) -> Literal[1]: ...
    @overload
    def __pow__(self, __x: _PositiveInteger, __modulo: None = ...) -> int: ...
    @overload
    def __pow__(self, __x: _NegativeInteger, __modulo: None = ...) -> float: ...
    # positive x -> int; negative x -> float
    # return type must be Any as `int | float` causes too many false-positive errors
    @overload
    def __pow__(self, __x: int, __modulo: None = ...) -> Any: ...
```
I'm slightly inclined against using an incompatible fallback overload.

Reply:
> I'm slightly inclined against using an incompatible fallback overload.

IMHO, because we can look at 2 type checkers, and one of them (`mypy`) seems to report non-existent issues, I think we should add the extra overloads to make `mypy` happy, as long as `pyright` wouldn't identify that as problematic.

Reply:
I reverted the overlapping overloads: while pyright and mypy correctly infer the types in most cases, both report errors (need to enable reportOverlappingOverload for pyright):

```py
# pyright: strict
from __future__ import annotations

from typing import Literal, overload

@overload
def isna(x: int) -> bool:
    ...

@overload
def isna(x: object) -> Literal[False]:
    ...

```    
    
I think the path to more fine-grained overloads is, as already started by @jbrockmendel, to lobby for something like NotType[...] to create non-overlapping overloads.

Reply:
FWIW i weighed in on a previously existing thread asking for examples when a NotType would be useful, and Guido offered a solution to my use case that didn't use NotType.  i.e. contributing to that thread would probably be worthwhile if you want to see this happen